test_permission_request_net is flaky on macOS

example: https://github.com/denoland/deno/runs/460692017

```
2020-02-21T17:12:43.4909520Z test worker::tests::execute_mod_resolve_error ... ok
2020-02-21T17:12:43.5254560Z test permissions::tests::test_permission_request_net ... FAILED
2020-02-21T17:12:43.6086890Z [1;38;5;10mDownload[0m http://localhost:4545/cli/tests/subdir/mod2.ts
2020-02-21T17:12:43.6121720Z [1;38;5;10mDownload[0m http://localhost:4545/cli/tests/subdir/print_hello.ts
2020-02-21T17:12:44.4458190Z test worker::tests::execute_006_url_imports ... ok
2020-02-21T17:12:44.4458680Z 
2020-02-21T17:12:44.4459260Z failures:
2020-02-21T17:12:44.4459750Z 
2020-02-21T17:12:44.4461350Z ---- permissions::tests::test_permission_request_net stdout ----
2020-02-21T17:12:44.4462430Z thread 'permissions::tests::test_permission_request_net' panicked at 'assertion failed: `(left == right)`
2020-02-21T17:12:44.4462630Z   left: `Deny`,
2020-02-21T17:12:44.4463200Z  right: `Allow`', cli/permissions.rs:630:5
2020-02-21T17:12:44.4463380Z stack backtrace:
2020-02-21T17:12:44.4464300Z    0:        0x1095cf8b5 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h109831b4e69a3cdc
2020-02-21T17:12:44.4465160Z    1:        0x1095f73d0 - core::fmt::write::haefb4093ced4abcf
2020-02-21T17:12:44.4466020Z    2:        0x10862cf49 - std::io::Write::write_fmt::h57d9d260264ecc5b
2020-02-21T17:12:44.4466940Z    3:        0x1095c5cdc - std::io::impls::<impl std::io::Write for alloc::boxed::Box<W>>::write_fmt::hb0f0233a36cd2ce6
2020-02-21T17:12:44.4468400Z    4:        0x1095d1da3 - std::panicking::default_hook::{{closure}}::hd11104060d683e7f
2020-02-21T17:12:44.4469350Z    5:        0x1095d1a57 - std::panicking::default_hook::h3ac89e0bbec67847
2020-02-21T17:12:44.4470230Z    6:        0x1095d244b - std::panicking::rust_panic_with_hook::h497844b5b7d1708e
2020-02-21T17:12:44.4470760Z    7:        0x1095d1fe9 - rust_begin_unwind
2020-02-21T17:12:44.4471610Z    8:        0x109618e4f - std::panicking::begin_panic_fmt::hc560df0756655667
2020-02-21T17:12:44.4472500Z    9:        0x1085159ba - deno::permissions::tests::test_permission_request_net::hb73241253f7adf15
2020-02-21T17:12:44.4473430Z   10:        0x10863808e - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h44c450ea090845aa
2020-02-21T17:12:44.4473980Z   11:        0x1095dc50f - __rust_maybe_catch_panic
2020-02-21T17:12:44.4474840Z   12:        0x108651955 - test::run_test::run_test_inner::{{closure}}::h433eeed657bd4954
2020-02-21T17:12:44.4476080Z   13:        0x10862c40b - std::sys_common::backtrace::__rust_begin_short_backtrace::h06b79d7b3e88f6bd
2020-02-21T17:12:44.4476960Z   14:        0x1086308db - std::panicking::try::do_call::hcb5b9f07da0d5a0c
2020-02-21T17:12:44.4477500Z   15:        0x1095dc50f - __rust_maybe_catch_panic
2020-02-21T17:12:44.4478380Z   16:        0x1086312b6 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h01ec0e0a5cf97627
2020-02-21T17:12:44.4479920Z   17:        0x1095bff1e - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hf89a561d6d706029
2020-02-21T17:12:44.4480880Z   18:        0x1095db96e - std::sys::unix::thread::Thread::new::thread_start::hd09405f1ea9f0a0f
2020-02-21T17:12:44.4481540Z   19:     0x7fff6c4e0e65 - _pthread_start
2020-02-21T17:12:44.4481690Z 
2020-02-21T17:12:44.4481750Z 
2020-02-21T17:12:44.4481900Z failures:
2020-02-21T17:12:44.4482080Z     permissions::tests::test_permission_request_net
2020-02-21T17:12:44.4482190Z 
2020-02-21T17:12:44.4482380Z test result: FAILED. 200 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
2020-02-21T17:12:44.4482470Z 
2020-02-21T17:12:44.4487490Z error: test failed, to rerun pass '-p deno --lib'
2020-02-21T17:12:44.4509090Z ##[error]Process completed with exit code 101.
2020-02-21T17:12:44.5041690Z Cleaning up orphan processes
2020-02-21T17:12:44.6483790Z Terminate orphan process: pid (1249) (sccache)
```

--- a/cli/permissions.rs
+++ b/cli/permissions.rs
@@ -15,6 +15,8 @@ use std::path::{Path, PathBuf};
 use std::sync::atomic::AtomicBool;
 #[cfg(test)]
 use std::sync::atomic::Ordering;
+#[cfg(test)]
+use std::sync::Mutex;
 use url::Url;
 
 const PERMISSION_EMOJI: &str = "⚠️";
@@ -324,6 +326,12 @@ fn permission_prompt(message: &str) -> bool {
 }
 
 #[cfg(test)]
+lazy_static! {
+  /// Lock this when you use `set_prompt_result` in a test case.
+  static ref PERMISSION_PROMPT_GUARD: Mutex<()> = Mutex::new(());
+}
+
+#[cfg(test)]
 static STUB_PROMPT_VALUE: AtomicBool = AtomicBool::new(true);
 
 #[cfg(test)]
@@ -519,6 +527,7 @@ mod tests {
 
   #[test]
   fn test_permissions_request_run() {
+    let guard = PERMISSION_PROMPT_GUARD.lock().unwrap();
     let mut perms0 = DenoPermissions::from_flags(&DenoFlags {
       ..Default::default()
     });
@@ -530,10 +539,12 @@ mod tests {
     });
     set_prompt_result(false);
     assert_eq!(perms1.request_run(), PermissionState::Deny);
+    drop(guard);
   }
 
   #[test]
   fn test_permissions_request_read() {
+    let guard = PERMISSION_PROMPT_GUARD.lock().unwrap();
     let whitelist = vec![PathBuf::from("/foo/bar")];
     let mut perms0 = DenoPermissions::from_flags(&DenoFlags {
       read_whitelist: whitelist.clone(),
@@ -566,10 +577,12 @@ mod tests {
       perms2.request_read(&Some(Path::new("/foo/baz"))),
       PermissionState::Deny
     );
+    drop(guard);
   }
 
   #[test]
   fn test_permissions_request_write() {
+    let guard = PERMISSION_PROMPT_GUARD.lock().unwrap();
     let whitelist = vec![PathBuf::from("/foo/bar")];
     let mut perms0 = DenoPermissions::from_flags(&DenoFlags {
       write_whitelist: whitelist.clone(),
@@ -602,10 +615,12 @@ mod tests {
       perms2.request_write(&Some(Path::new("/foo/baz"))),
       PermissionState::Deny
     );
+    drop(guard);
   }
 
   #[test]
   fn test_permission_request_net() {
+    let guard = PERMISSION_PROMPT_GUARD.lock().unwrap();
     let whitelist = svec!["localhost:8080"];
 
     let mut perms0 = DenoPermissions::from_flags(&DenoFlags {
@@ -652,10 +667,12 @@ mod tests {
     });
     set_prompt_result(true);
     assert!(perms3.request_net(&Some(":")).is_err());
+    drop(guard);
   }
 
   #[test]
   fn test_permissions_request_env() {
+    let guard = PERMISSION_PROMPT_GUARD.lock().unwrap();
     let mut perms0 = DenoPermissions::from_flags(&DenoFlags {
       ..Default::default()
     });
@@ -667,10 +684,12 @@ mod tests {
     });
     set_prompt_result(false);
     assert_eq!(perms1.request_env(), PermissionState::Deny);
+    drop(guard);
   }
 
   #[test]
   fn test_permissions_request_plugin() {
+    let guard = PERMISSION_PROMPT_GUARD.lock().unwrap();
     let mut perms0 = DenoPermissions::from_flags(&DenoFlags {
       ..Default::default()
     });
@@ -682,10 +701,12 @@ mod tests {
     });
     set_prompt_result(false);
     assert_eq!(perms1.request_plugin(), PermissionState::Deny);
+    drop(guard);
   }
 
   #[test]
   fn test_permissions_request_hrtime() {
+    let guard = PERMISSION_PROMPT_GUARD.lock().unwrap();
     let mut perms0 = DenoPermissions::from_flags(&DenoFlags {
       ..Default::default()
     });
@@ -697,5 +718,6 @@ mod tests {
     });
     set_prompt_result(false);
     assert_eq!(perms1.request_hrtime(), PermissionState::Deny);
+    drop(guard);
   }
 }

